rfcs/109-redesign-features.md

```diff
@@ -0,0 +1,103 @@
+---
+author: Xuanwo <github@xuanwo.io>
+status: draft
+updated_at: 2021-06-17
+---
+
+# GSP-109: Redesign Features
+
+## Background
+
+[GSP-87](https://github.com/beyondstorage/specs/pull/87) introduces a new concept `Feature`. However, it's proved to be too complicated that both hard to understand and to implemented.
+
+In GSP-87, we added new struct `ServiceFeatures` and `StorageFeatures`:
+
+```go
+type ServiceFeatures struct {
+   LooseOperationAll    bool
+   LooseOperationCreate bool
+   LooseOperationDelete bool
+   LooseOperationGet    bool
+   LooseOperationList   bool
+
+   VirtualOperationAll bool
+
+   VirtualPairAll bool
+}
+
+type StorageFeatures struct {
+   LooseOperationAll               bool
+   LooseOperationCompleteMultipart bool
+   LooseOperationCreate            bool
+   LooseOperationCreateMultipart   bool
+   LooseOperationDelete            bool
+   LooseOperationList              bool
+   LooseOperationListMultipart     bool
+   LooseOperationMetadata          bool
+   LooseOperationRead              bool
+   LooseOperationStat              bool
+   LooseOperationWrite             bool
+   LooseOperationWriteMultipart    bool
+
+   VirtualOperationAll bool
+
+   VirtualPairAll bool
+}
+```
+
+There are the following problems:
+
+- `LooseOperationXxx` is too verbose and there is no user who requires this feature.
+- `VirtualOperationWriteDir` make developer-only care about it in `CreateDir`, but it's not correct: to implement `VirtualDir` support, the developer should check them everywhere.
+
+## Proposal
+
+Define features one by one and remove all autogenerated check, that means service implementer should check all features manually when required.
+
+For now, we will introduce the following features, other features should be introduced by new GSPs.
+
+### LoosePair
+
+We will introduce a new feature called `LoosePair` to replace `LooseOperation` features.
+
+If this feature is enabled, the service will not return an error for not support pairs.
+
+### VirtualDir
+
+We will introduce a new feature called `VirtualDir` to Replace `VirtualOperationCreateDir`.
+
+Any operation that is related to dir SHOULD check `VirtualDir` in services that don't have native support for dir.
+
+- `stat`: SHOULD check `VirtualDir` before set Dir ObjectMode
+- `create`: SHOULD check `VirtualDir` before accept Dir ObjectMode
+- `list`: SHOULD check `VirtualDir` before set Dir ObjectMode
+- `delete`: SHOULD check `VirtualDir` before delete a dir Object.
+- ...
+
+### VirtualObjectMetadata
+
+We will introduce a new feature called `VirtualObjectMetadata` to allow the service to handle object metadata even when it doesn't have native support.
```

Comment:
Why do we need to add this?

Reply:
Allow fs to use `xattr` to store object metadata like `content-type` and `content-md5`.

Reply:
I mean how do we do this before this proposal? I'm a little bit confused why it appears here.

Reply:
Is this refactoring or adding a new feature?

Reply:
> I mean how do we do this before this proposal? I'm a little bit confused why it appears here.

It's used to be added as `VirutalPairWriteContentMd5` / `VirtualPairWriteContentType` but never implemented.

In this proposal, we unify all of them into `VirtualObjectMetadata`.

Reply:
Got it. Can you add this explanation to the proposal?